src/models/token.rs

```diff
@@ -95,6 +97,30 @@ impl ApiToken {
         .or_else(|_| tokens.select(ApiToken::as_select()).first(conn))
         .map_err(Into::into)
     }
+
+    /// Find all tokens that are not revoked and will expire within the specified number of days.
+    pub fn find_tokens_expiring_within_days(
+        conn: &mut PgConnection,
+        days_until_expiry: i64,
```

Comment:
I think it might be better to accept a `chrono::TimeDelta` (aka. `chrono::Duration`) here to have a more expressive API.

Reply:
I refactored it in [`53f70bf` (#8290)](https://github.com/rust-lang/crates.io/pull/8290/commits/53f70bf814fe606526d4ca14c743f9efc04801ac)

But I'm not sure I used the right API. I used `now + days_until_expiry.num_days().day()` to check it. Do we have a better way to construct it?

Reply:
Since we're using `chrono::TimeDelta`, it might be more reasonable to calculate the actual time directly on the app side. This approach is similar to what we do in https://github.com/rust-lang/crates.io/blob/63c30c6cca0cfca1d3ad9be4816a43d3363c8b55/src/worker/jobs/downloads/clean_processed_log_files.rs#L31-L40.

Reply:
I tried to follow this approach. But I didn't make it compile. I will have another try. Thanks!

Reply:
@eth3lbert Does this make sense? We cannot directly use `chrono::DateTime<chrono::Utc>` here. Because expired_at is a Timestamp, not a Timestamptz.

```rust
    let expire_data = chrono::Utc::now() + days_until_expiry;
    api_tokens::table
        .filter(api_tokens::revoked.eq(false))
        .filter(
            api_tokens::expired_at
                .is_not_null()
                .and(api_tokens::expired_at.assume_not_null().gt(now)) // Ignore already expired tokens
                .and(
                    api_tokens::expired_at
                        .assume_not_null()
                        .lt(expire_data.naive_utc()),
                ),
        )
        .filter(api_tokens::expiry_notification_at.is_null())
        .select(ApiToken::as_select())
        .get_results(conn)
```

And I found we use my current approach in our codebase https://github.com/rust-lang/crates.io/blob/81023270df96b4d0f6cf78d0b41b4a7060737d74/src/controllers/token.rs#L49

Reply:
So I am not sure which one is better. It seems the current one is easier to understand and we don't need to call `naive_utc`.

Reply:
> Because expired_at is a Timestamp

Yeah, you need to use `NaiveDateTime`.
Ref: https://github.com/diesel-rs/diesel/blob/974814c6792c534736498a0bd7abf95e21f8bf7b/diesel/src/pg/types/date_and_time/chrono.rs#L21-L48

```rust
        let now = chrono::Utc::now();
        let expired_at = now + chrono::TimeDelta::try_days(days_until_expiry).unwrap();

        api_tokens::table
            .filter(api_tokens::revoked.eq(false))
            .filter(
                api_tokens::expired_at
                    .is_not_null()
                    .and(api_tokens::expired_at.assume_not_null().gt(now.naive_utc())) // Ignore already expired tokens
                    .and(
                        api_tokens::expired_at
                            .assume_not_null()
                            .lt(expired_at.naive_utc()),
                    ),
            )
            .filter(api_tokens::expiry_notification_at.is_null())
            .select(ApiToken::as_select())
            .get_results(conn)
```

Reply:
> So I am not sure which one is better. It seems the current one is easier to understand and we don't need to call `naive_utc`.

Either approach is fine with me. I actually left a similar comment before https://github.com/rust-lang/crates.io/pull/8278#discussion_r1520278505. 

One reason to use the app-side timestamp might be that readers wouldn't need to know how to use PG `intervals` to work with the data.

Reply:
> Yeah, you need to use `NaiveDateTime`.

Yes. I already did that. I think it makes the code more complex. Seems `now + days_until_expiry.num_days().day()` is easier to understand. I am not sure how PG does the query plan optimization for different approaches. Usually, it shouldn't be a big deal, and also to evaluate the value of this simple expression won't be a big deal. 

So I prefer to keep `now + days_until_expiry.num_days().day()` to keep it simple. 

But if it has a huge performance difference, then maybe we should change it. 

Thanks!